### PR TITLE
Revert statement.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 1]
```diff
@@ -281,42 +291,59 @@ struct ModifierContextChecker: public PostTypeChecker::Checker
...
-				);
+	void endVisit(RevertStatement const&) override
+	{
+		m_currentStatement = nullptr;
+	}
+
+	bool visit(FunctionCall const& _functionCall) override
+	{
+		if (*_functionCall.annotation().kind == FunctionCallKind::FunctionCall)
+			if (FunctionTypePointer const functionType = dynamic_cast<FunctionTypePointer const>(_functionCall.expression().annotation().type))
```
Comment (mijovic) [inline libsolidity/analysis/PostTypeChecker.cpp]: Maybe?
```suggestion
			if (auto const* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type))
```

File: libsolidity/ast/AST.h [hunk 2]
```diff
@@ -1722,6 +1722,28 @@ class Throw: public Statement
...
+class RevertStatement: public Statement
+{
+public:
+	explicit RevertStatement(
+		int64_t _id,
+		SourceLocation const& _location,
+		ASTPointer<ASTString> const& _docString,
+		ASTPointer<FunctionCall> _functionCall
+	):
+		Statement(_id, _location, _docString), m_errorCall(std::move(_functionCall)) {}
```
Comment (mijovic) [inline libsolidity/ast/AST.h]: What about adding assertion?
```suggestion
		Statement(_id, _location, _docString), m_errorCall(std::move(_functionCall))
	{
		solAssert(m_errorCall, "");
	}
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -1043,7 +1043,29 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 	}
 	case FunctionType::Kind::Error:
 	{
-		solAssert(false, "");
+		vector<string> arguments;
+		vector<Type const*> argumentTypes;
+		for (ASTPointer<Expression const> const& arg: _functionCall.sortedArguments())
+		{
+			argumentTypes.push_back(arg->annotation().type);
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Should we add assertion before this call?
`solAssert(!arg->annotation().type, "");`

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 4]
```diff
@@ -1043,7 +1043,29 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+			arguments += IRVariable(*arg).stackSlots();
+		}
+		ABIFunctions abi(m_context.evmVersion(), m_context.revertStrings(), m_context.functionCollector());
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Maybe moving it down when it is used. Also, you can inline it with using `m_context.abiFunctions()`

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -1966,6 +1988,13 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 					);
 						// the call will do the resolving
 					break;
+				case FunctionType::Kind::Error:
+					solAssert(
+						dynamic_cast<ErrorDefinition const*>(_memberAccess.annotation().referencedDeclaration),
+						"Error not found"
+					);
+						// The function call will resolve the selector.
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
					// The function call will resolve the selector.
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 6]
```diff
@@ -1043,7 +1043,29 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 	}
 	case FunctionType::Kind::Error:
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1046]: Should we just call `revertWithError` here?

**Review by mijovic — COMMENTED** [review]
Few small comments and suggestions for first round.

Also this needs to be rebased again, as there are some commits from other PRs

File: libsolidity/ast/AST.h (hunk 2 above)
Comment (chriseth) [reply libsolidity/ast/AST.h]: We almost never do that in the AST, but I can add it.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 6 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1046]: Interesting! I think I would rather simplify `revertWithError` for now.

Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1046]: I just asked as it was looking almost the same like this block here. 
So simplifying sounds good to me

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1046]: I actually ended up calling `revertWithError` :)

File: libsolidity/ast/AST.cpp [hunk 7]
```diff
@@ -165,9 +166,14 @@ vector<EventDefinition const*> const& ContractDefinition::interfaceEvents() cons
 vector<ErrorDefinition const*> ContractDefinition::interfaceErrors() const
 {
 	set<ErrorDefinition const*, CompareByID> result;
-	// TODO add all referenced errors
 	for (ContractDefinition const* contract: annotation().linearizedBaseContracts)
 		result += filteredNodes<ErrorDefinition>(contract->m_subNodes);
+	solAssert(annotation().creationCallGraph.set() == annotation().deployedCallGraph.set(), "");
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp:183]: Now we are collecting all errors that are referenced. The problem is that this function is used too early in the "hash collision checker", at a point where we do not even have the call graph yet.

File: libsolidity/analysis/PostTypeContractLevelChecker.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+			noErrors = false;
+	return noErrors;
+}
+
+bool PostTypeContractLevelChecker::check(ContractDefinition const& _contract)
+{
+	solAssert(
+		_contract.annotation().creationCallGraph.set() &&
+		_contract.annotation().deployedCallGraph.set(),
+	"" );
```
Comment (cameel) [inline libsolidity/analysis/PostTypeContractLevelChecker.cpp]:
```suggestion
	solAssert(
		_contract.annotation().creationCallGraph.set() &&
		_contract.annotation().deployedCallGraph.set(),
		""
	);
```

File: libsolidity/analysis/PostTypeContractLevelChecker.cpp [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+		string signature = error->functionType(true)->externalSignature();
+		uint32_t hash = selectorFromSignature32(signature);
+		if (!errorHashes[hash].empty() && !errorHashes[hash].count(signature))
+		{
+			SourceLocation& otherLocation = errorHashes[hash].begin()->second;
+			m_errorReporter.typeError(
+				4883_error,
+				error->nameLocation(),
+				SecondarySourceLocation{}.append("This error has a different signature but the same hash: ", otherLocation),
+				string("Error signature hash collision for ") + error->functionType(true)->externalSignature()
```
Comment (cameel) [inline libsolidity/analysis/PostTypeContractLevelChecker.cpp]: These two suggestions from the syntax PR also apply here: https://github.com/ethereum/solidity/pull/10987#discussion_r596822244, https://github.com/ethereum/solidity/pull/10987#discussion_r596827761

File: libsolidity/analysis/PostTypeContractLevelChecker.cpp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+		_contract.annotation().creationCallGraph.set() &&
+		_contract.annotation().deployedCallGraph.set(),
+	"" );
+
+	map<uint32_t, map<string, SourceLocation>> errorHashes;
+	for (ErrorDefinition const* error: _contract.interfaceErrors())
+	{
+		string signature = error->functionType(true)->externalSignature();
+		uint32_t hash = selectorFromSignature32(signature);
+		if (!errorHashes[hash].empty() && !errorHashes[hash].count(signature))
```
Comment (cameel) [inline libsolidity/analysis/PostTypeContractLevelChecker.cpp:57]: I'd suggest `valueOrDefault()` here (though we still have the unresolved lifetime problem: https://github.com/ethereum/solidity/pull/10973#discussion_r580926054):

```suggestion
		if (valueOrDefault(errorHashes, hash, {}).count(signature) == 0)
```

File: libsolidity/ast/AST.cpp [hunk 11]
```diff
@@ -165,9 +166,14 @@ vector<EventDefinition const*> const& ContractDefinition::interfaceEvents() cons
...
+	if (annotation().creationCallGraph.set())
+	{
+		result += (*annotation().creationCallGraph)->usedErrors;
+		result += (*annotation().deployedCallGraph)->usedErrors;
+	}
 	return convertContainer<vector<ErrorDefinition const*>>(move(result));
 }
```
Comment (cameel) [inline libsolidity/ast/AST.cpp]: I think it's rather error prone that this function could just silently not give you all errors if you call it too early in the analysis.

How about adding a parameter that states whether the caller needs the errors from the graph or not and having an assertion that fails if he does and they are not yet available?

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 12]
```diff
@@ -1832,6 +1846,7 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 		solAssert(
 			dynamic_cast<VariableDeclaration const*>(_memberAccess.annotation().referencedDeclaration) ||
 			dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration) ||
+			dynamic_cast<ErrorDefinition const*>(_memberAccess.annotation().referencedDeclaration) ||
```
Comment (cameel) [inline libsolidity/codegen/ExpressionCompiler.cpp:1885]: Don't we also need to handle `Type::Category::Type` in this function?

File: libsolidity/formal/SMTEncoder.cpp [hunk 13]
```diff
@@ -2477,7 +2456,23 @@ void SMTEncoder::defineExpr(Expression const& _e, smtutil::Expression _value)
...
+	}
+	auto const& symbTuple = dynamic_pointer_cast<smt::SymbolicTupleVariable>(m_context.expression(_e));
+	solAssert(symbTuple, "");
+	symbTuple->increaseIndex();
+	auto const& symbComponents = symbTuple->components();
+	solAssert(symbComponents.size() == _values.size(), "");
+	for (unsigned i = 0; i < symbComponents.size(); ++i)
+		if (_values[i])
+			m_context.addAssertion(symbTuple->component(i) == *_values[i]);
 }
```
Comment (cameel) [inline libsolidity/formal/SMTEncoder.cpp]: Looks like this is already in `develop` so this PR probably needs a rebase. Seems unrelated to the revert statement.

File: test/libsolidity/syntaxTests/revertStatement/regular.sol [hunk 14]
```diff
@@ -0,0 +1,6 @@
+error E();
+contract C {
+    function f() public pure {
+        revert E();
+    }
+}
```
Comment (cameel) [inline test/libsolidity/syntaxTests/revertStatement/regular.sol:6]: Some more syntax test ideas (just listing anything I think could potentially be an edge case, please ignore the ones that you think do not add any value or are already covered elsewhere):
- `revert revert;`
- `error revert(string);`
- `function revert() {}`
- `revert E;`
- `revert Event();` (where `Event` is an event)
- No reported hash collision when you only refer to the error by name (without "calling" it).
- No reported hash collision when you only use `E.selector`.
- `revert Panic();` - should be disallowed if not shadowed by another, renamed error, right?

File: test/libsolidity/semanticTests/errors/panic_via_import.sol [hunk 15]
```diff
@@ -0,0 +1,13 @@
+==== Source: s1.sol ====
+error E(uint);
+==== Source: s2.sol ====
+import { E as Panic } from "s1.sol";
+contract C {
+  function a() public pure {
+    revert Panic(1);
+  }
```
Comment (cameel) [inline test/libsolidity/semanticTests/errors/panic_via_import.sol]:
```suggestion
    function a() public pure {
        revert Panic(1);
    }
```

By the way, didn't we want to disallow shadowing `Panic` and `Error`?

File: test/libsolidity/semanticTests/errors/using_structs.sol [hunk 16]
```diff
@@ -0,0 +1,16 @@
+pragma abicoder v2;
+struct S { uint a; string b; }
+error E(uint a, S, uint b);
+contract C {
+    S s;
+    function f(bool c) public {
+        s.a = 9;
+        s.b = "abc";
+        revert E(2, s, 7);
+    }
```
Comment (cameel) [inline test/libsolidity/semanticTests/errors/using_structs.sol]: Why the unused parameter? Was the original intention to use it in the struct?

**Review by cameel — COMMENTED** [review]
Only minor suggestions. I haven't found anything serious.

File: libsolidity/analysis/PostTypeContractLevelChecker.cpp (hunk 9 above)
Comment (chriseth) [reply libsolidity/analysis/PostTypeContractLevelChecker.cpp]: The second does not work for me.

File: libsolidity/analysis/PostTypeContractLevelChecker.cpp (hunk 10 above)
Comment (chriseth) [reply libsolidity/analysis/PostTypeContractLevelChecker.cpp:57]: I think your suggestion changes the semantics.
The condition is: There is a different signature for the same hash in the map. I will add a comment.

File: libsolidity/ast/AST.cpp (hunk 11 above)
Comment (chriseth) [reply libsolidity/ast/AST.cpp]: Yes, I will add that.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 12 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:1885]: There is no `Type::Category::Type`, but `TypeType` is handled one line further down...

File: test/libsolidity/semanticTests/errors/panic_via_import.sol (hunk 15 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/errors/panic_via_import.sol]: They are not global variables, so we cannot really do shadowing there.

File: test/libsolidity/semanticTests/errors/using_structs.sol (hunk 16 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/errors/using_structs.sol]: Good question! Maybe I wanted to one test via memory and one via storage. Will change it accordingly.

File: libsolidity/analysis/PostTypeContractLevelChecker.cpp (hunk 10 above)
Comment (cameel) [reply libsolidity/analysis/PostTypeContractLevelChecker.cpp:57]: Sorry, I must have just skimmed over it and misintepreted it as something where you first check if the container exists and only then do an operation on it.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 12 above)
Comment (cameel) [reply libsolidity/codegen/ExpressionCompiler.cpp:1885]: Right, I meant `TypeType`.

Well, I should have been more precise. I was asking about the category of the expression which is what the `switch` checks, i.e. `_memberAccess.expression().annotation().type->category() == Type::Category::TypeType`.

But ok, I think I have an answer. It's handled separately before the `switch`.

**Review by cameel — DISMISSED** [review]
LTGM

File: libsolidity/interface/CompilerStack.cpp [hunk 17]
```diff
@@ -433,6 +434,14 @@ bool CompilerStack::analyze()
 				noErrors = false;
 		}
 
+		if (noErrors)
+		{
```
Comment (mijovic) [inline libsolidity/interface/CompilerStack.cpp]: I think brackets are not needed, but not mandatory to change

**Review by mijovic — APPROVED** [review]
Looks good.

**Comment by hrkrshnn** [comment]
Can you also add this test:

```
struct error { uint error; }
contract C {
	uint error = 0;
	error test();
	function f() public {
		revert test();
	}
}
contract D {
	error _struct;
	error test();
}
```

should succeed (already does) because `error` is not a keyword.

**Comment by chriseth** [comment]
It actually fails :(

**Comment by chriseth** [comment]
Added the test as `error_structs.sol`.

**Comment by chriseth** [comment]
Rebased without changes.

**Review by mijovic — DISMISSED** [review]

File: libsolidity/analysis/FunctionCallGraph.cpp [hunk 18]
```diff
@@ -125,6 +125,8 @@ bool FunctionCallGraphBuilder::visit(FunctionCall const& _functionCall)
 		// change at runtime). All we can do is to add an edge to the dispatch which in turn has
 		// edges to all functions could possibly be called.
 		add(m_currentNode, CallGraph::SpecialNode::InternalDispatch);
+	else if (functionType->kind() == FunctionType::Kind::Error)
+		m_graph.usedErrors.insert(&dynamic_cast<ErrorDefinition const&>(functionType->declaration()));
```
Comment (aarlt) [inline libsolidity/analysis/FunctionCallGraph.cpp:129]: maybe we should add an assert here?

Comment (cameel) [reply libsolidity/analysis/FunctionCallGraph.cpp:129]: You mean as `else`? Not really. There are other case possible here, the graph just skips them.

Comment (chriseth) [reply libsolidity/analysis/FunctionCallGraph.cpp:129]: There is a check and I would say it is highly unlikely for this assert to trigger...

**Comment by hrkrshnn** [comment]
```
  struct S {uint x;}
  contract C {
	  function f() public {
		revert S(10);
	  }
  }
```
produces an ICE:

```
: Internal compiler error during compilation:
: /solidity/libsolidity/codegen/ContractCompiler.cpp(79): Throw in function void {anonymous}::StackHeightChecker::check()
: Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
: std::exception::what: I sense a disturbance in the stack: 2 vs 1
: [solidity::util::tag_comment*] = I sense a disturbance in the stack: 2 vs 1
```

But the following works!

```
  struct S {uint x;}
  contract C {
	  error S(uint y);
	  function f() public {
		revert S(10);
	  }
  }
```

File: test/libsolidity/syntaxTests/revertStatement/revert_event.sol [hunk 19]
```diff
@@ -0,0 +1,8 @@
+contract C {
+    event E();
+    function f() public pure {
+        revert E();
```
Comment (aarlt) [inline test/libsolidity/syntaxTests/revertStatement/revert_event.sol:4]: I was wondering why pure functions are allowed to revert. I think that should be forbidden.

Comment (chriseth) [reply test/libsolidity/syntaxTests/revertStatement/revert_event.sol:4]: This has been discussed a lot in the past. Pure means the result only depends on the arguments, and that is the case here.

**Comment by chriseth** [comment]
@hrkrshnn oh wow! Good that you found this! I could swear the code for that was there before...

**Comment by chriseth** [comment]
@hrkrshnn added the check to TypeChecker and added tests.
